File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-java/src/main/java/com/octocat/supply/controller/DeliveriesController.java
package com.octocat.supply.controller;

import com.octocat.supply.exception.NotFoundException;
import com.octocat.supply.model.DeliveryModel.CreateDeliveryRequest;
import com.octocat.supply.model.DeliveryModel.Delivery;
import com.octocat.supply.model.DeliveryModel.UpdateDeliveryRequest;
import com.octocat.supply.model.DeliveryModel.UpdateDeliveryStatusRequest;
import com.octocat.supply.repository.DeliveriesRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/deliveries")
public class DeliveriesController {
    private final DeliveriesRepository repository;

    public DeliveriesController(DeliveriesRepository repository) {
        this.repository = repository;
    }

    @GetMapping
    public List<Delivery> getAll() {
        return repository.findAll();
    }

    @GetMapping("/{id}")
    public Delivery getById(@PathVariable int id) {
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Delivery not found"));
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Delivery create(@RequestBody CreateDeliveryRequest request) {
        ControllerValidation.requireText(request.name(), "name");
        ControllerValidation.requireText(request.deliveryDate(), "deliveryDate");
        return repository.create(request);
    }

    @PutMapping("/{id}")
    public Delivery update(@PathVariable int id, @RequestBody UpdateDeliveryRequest request) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("Delivery not found"));
        repository.update(id, request);
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Delivery not found"));
    }

    @PutMapping("/{id}/status")
    public Object updateStatus(@PathVariable int id, @RequestBody UpdateDeliveryStatusRequest request) {
        ControllerValidation.requireText(request.status(), "status");
        repository.findById(id).orElseThrow(() -> new NotFoundException("Delivery not found"));
        repository.updateStatus(id, request.status());
        Delivery updated = repository.findById(id).orElseThrow(() -> new NotFoundException("Delivery not found"));
        if (request.deliveryPartner() != null && !request.deliveryPartner().isBlank()) {
            Map<String, Object> response = new LinkedHashMap<>();
            response.put("delivery", updated);
            response.put("commandOutput", "");
            return response;
        }
        return updated;
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable int id) {
        if (!repository.delete(id)) {
            throw new NotFoundException("Delivery not found");
        }
    }
}

File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-cs/Controllers/DeliveriesController.cs
using Microsoft.AspNetCore.Mvc;
using OctoSupply.Api.Models;
using OctoSupply.Api.Repositories;
using OctoSupply.Api.Utils;

namespace OctoSupply.Api.Controllers;

[ApiController]
[Route("api/deliveries")]
public sealed class DeliveriesController(DeliveriesRepository repository) : ControllerBase
{
    private readonly DeliveriesRepository _repository = repository;

    public sealed class UpdateDeliveryStatusRequest
    {
        public string Status { get; set; } = string.Empty;
        public string? DeliveryPartner { get; set; }
    }

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<Delivery>>> GetAll() =>
        Ok(await _repository.FindAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<Delivery>> GetById(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        return item is null ? throw new NotFoundException("Delivery", id) : Ok(item);
    }

    [HttpPost]
    public async Task<ActionResult<Delivery>> Create([FromBody] CreateDeliveryRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Name))
        {
            throw new ValidationException("name is required");
        }

        if (string.IsNullOrWhiteSpace(request.DeliveryDate))
        {
            throw new ValidationException("deliveryDate is required");
        }

        var created = await _repository.CreateAsync(request);
        return CreatedAtAction(nameof(GetById), new { id = created.DeliveryId }, created);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<Delivery>> Update(int id, [FromBody] UpdateDeliveryRequest request) =>
        Ok(await _repository.UpdateAsync(id, request));

    [HttpPut("{id:int}/status")]
    public async Task<ActionResult<Delivery>> UpdateStatus(int id, [FromBody] UpdateDeliveryStatusRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Status))
        {
            throw new ValidationException("status is required");
        }

        var updated = await _repository.UpdateStatusAsync(id, request.Status);
        if (!string.IsNullOrWhiteSpace(request.DeliveryPartner))
        {
            return Ok(new { delivery = updated, commandOutput = string.Empty });
        }

        return Ok(updated);
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        await _repository.DeleteAsync(id);
        return NoContent();
    }
}
